### PR TITLE
53 signup login

### PR DESCRIPTION
**Pull request by decembrya**
### What does it fix?

Part of #53

Add endpoints to register, login and delete a user account along with Swagger for these for easy dev use.
Adding JWT Bearer authentication on GraphQL.
In order to allow easier setting of Auth header for devs added GraphQL Playground instead of GraphiQl

Still need to do in a separate PR: Google, FB auth 
### How has it been tested?

Tested manually, updated and ran available specs 
ran using docker-compose

### CODE REVIEW

**Comment by pirvudoru** [comment]
WOW

File: TeacherWorkout.Specs/Features/Themes.feature [hunk 1]
```diff
@@ -7,9 +7,9 @@ Scenario: Admin user can create a theme
 	When Ion creates a theme
 	Then the theme was created successfully
 
-Scenario: Anonymous user can list themes
+Scenario: Authenticated user can list themes
```
Comment (pirvudoru) [inline TeacherWorkout.Specs/Features/Themes.feature:10]: I think we should keep support for read operations public. I think the public website will need at least lessonsto be public. And probably the lesson will display somewhere the theme it is part of.

https://www.figma.com/file/uLiqrlxmOB5xCppIzmiUXV/Teacher-Workout?node-id=1%3A2

File: TeacherWorkout.Api/Startup.cs [hunk 2]
```diff
@@ -48,12 +56,27 @@ public void ConfigureServices(IServiceCollection services)
...
                 .AddErrorInfoProvider(opt => opt.ExposeExceptionStackTrace = true)
                 .AddSystemTextJson()
-                .AddGraphTypes();
+                .AddGraphTypes()
+                .AddUserContextBuilder(httpContext => new GraphQlUserContext(httpContext.User));
+            
+            services.AddControllers();
+
+            var applicationAssemblies = GetAssemblies();
+            services.AddSwaggerFor(applicationAssemblies, Configuration);
```
Comment (pirvudoru) [inline TeacherWorkout.Api/Startup.cs:65]: Can we add to the readme the URL to the swagger UI, next to the GraphiQL mention

File: TeacherWorkout.Specs/Features/Themes.feature (hunk 1 above)
Comment (pirvudoru) [reply TeacherWorkout.Specs/Features/Themes.feature:10]: We should add authorization for the mutations (lessonSave, stepComplete, themeCreate), and the lessonStatuses query since they rely on a user to save data against.

File: TeacherWorkout.Api/Startup.cs (hunk 2 above)
Comment (pirvudoru) [reply TeacherWorkout.Api/Startup.cs:65]: http://localhost:5000/swagger/index.html

File: TeacherWorkout.Api/Startup.cs [hunk 3]
```diff
@@ -48,12 +56,27 @@ public void ConfigureServices(IServiceCollection services)
...
 
             services.AddDbContext<TeacherWorkoutContext>(options =>
                 options.UseNpgsql(Configuration.GetConnectionString("TeacherWorkoutContext")));
+
+            services.AddIdentityApiServices(Configuration);
```
Comment (pirvudoru) [inline TeacherWorkout.Api/Startup.cs:70]: The initial discussion with @catileptic was that we should use IdentityServer (mentioned in the issue). From what I understand from the code, we are not using it. I am curious why the change.

File: TeacherWorkout.Api/TeacherWorkout.Api.csproj [hunk 4]
```diff
@@ -2,21 +2,24 @@
 
   <PropertyGroup>
     <TargetFramework>net5.0</TargetFramework>
+    <UserSecretsId>af66c3e1-ddfc-4130-9207-6a6bdc33931e</UserSecretsId>
   </PropertyGroup>
 
   <ItemGroup>
     <PackageReference Include="GraphQL.Server.Transports.AspNetCore.SystemTextJson" Version="5.0.2" />
-    <PackageReference Include="GraphQL.Server.Ui.Graphiql" Version="5.0.2" />
+    <PackageReference Include="GraphQL.Server.Ui.Playground" Version="5.2.0" />
```
Comment (pirvudoru) [inline TeacherWorkout.Api/TeacherWorkout.Api.csproj:10]: Please also update the readme to point to the graphql docs

http://localhost:5000/ui/playground

File: TeacherWorkout.Api/appsettings.Development.json [hunk 5]
```diff
@@ -8,6 +8,25 @@
   },
   "ConnectionStrings": {
     //connection string for local environment
-    "TeacherWorkoutContext": "host=localhost;Port=5432;Database=teacher_workout;User Id=docker;Password=docker;"
-  }
+    "TeacherWorkoutContext": "host=localhost;Port=5432;Database=teacher_workout;User Id=docker;Password=docker;",
+    "IdentityDbConnectionString": "host=localhost;Port=5432;Database=teacher_workout;User Id=docker;Password=docker;"
+  },
+  "JwtConfig": {
+    "Secret": "",
```
Comment (pirvudoru) [inline TeacherWorkout.Api/appsettings.Development.json]: We should add some values here or load from .env. I was getting errors when trying to start the project in development mode after pulling the branch locally.

File: README.md [hunk 6]
```diff
@@ -76,7 +76,12 @@ TODO: decouple docker-compose in development from staging
 
 ### Add a migration
 ```
-dotnet ef migrations add <MigrationNameGoesHere> --startup-project TeacherWorkout.Api/ --project TeacherWorkout.Data/
+dotnet ef migrations add <MigrationNameGoesHere> --startup-project TeacherWorkout.Api/ --project TeacherWorkout.Data/ --context TeacherWorkoutContext
+```
+
+### Add a migration for a different context
+```
+dotnet ef migrations add <MigrationNameGoesHere> --context UserContext --project .\TeacherWorkout.Identity --startup-project .\TeacherWorkout.Api
```
Comment (pirvudoru) [inline README.md:103]: Do we also need to run the migrations for the separate context?

File: .env.example [hunk 7]
```diff
@@ -7,4 +7,9 @@ EXPOSED_DB_PORT=5432
...
 POSTGRES_USER=docker
 POSTGRES_DB=teacher_workout
-SEEDDATA=false
\ No newline at end of file
+SEEDDATA=false
+SEEDUSERS=false
+ADMIN_EMAIL=
+ADMIN_PASSWORD=
+
+JWT_SECRET=
```
Comment (pirvudoru) [inline .env.example]: Can we add default values so that someone new to the project can just copy .env.example -> .env and is able to run the project?

File: TeacherWorkout.Api/appsettings.CI.json [hunk 8]
```diff
@@ -7,6 +7,25 @@
...
+    "RequireNonAlphanumeric": false,
+    "RequireLowercase": false,
+    "RequireUppercase": false,
+    "RequireDigit": false
+  },
+  "Admin": {
+    "EmailAddress": "",
+    "Password": ""
+  },
+  "SeedUsers": false,
```
Comment (pirvudoru) [inline TeacherWorkout.Api/appsettings.CI.json:29]: Why do we need the separation between Users and Data to seed ? I'd assume we just run the seed and it will take care of everything. If we do, please update the Readme to reference how we seed all the data (imagine you are cloning the project and trying to set it up from scratch).

**Review by pirvudoru — CHANGES_REQUESTED** [review]
Thank you for your contribution. It's huge.

Please have a look at the comments, and please rebase or pull develop.

**Comment by decembrya** [comment]
Closing so we can reimplement using identity server
